Viewport Fit Cover + Font Family Fallbacks

Offroaders123 opened the pull request:
I really like the design of your website! Thought it would look nice with fullscreen support for devices with camera cutouts, like modern iPhones and the like.

### Viewport Fit: Default
![Screenshot of Viewport Fit Default](https://user-images.githubusercontent.com/65947371/162845637-7e2744c3-881e-4081-865f-04f3edd6a438.PNG)

### Viewport Fit: Cover
![Screenshot of Viewport Fit Cover](https://user-images.githubusercontent.com/65947371/162845733-8da52517-e14a-42ac-817e-0c89e1fb69c7.PNG)

Here's a list of the things I changed in the commit:

Additions:
- Added fullscreen support for devices that take advantage of the Safe Area Inset environment variables in CSS. This allows the body's background image to take up all of the space on the screen, where available. This is enabled from within the viewport meta tag, which then allows the page layout to be fully managed from CSS after that.
- Added a few fallbacks to the body's font family declaration. I noticed that system-ui wasn't working on Chrome OS, and I believe Firefox still doesn't have support for it yet. These additional fallbacks will allow the system font to appear correctly for more devices and browsers that don't yet support system-ui.

Offroaders123 committed:
Viewport Fit Cover + Font Family Fallbacks

Additions:
- Added fullscreen support for devices that take advantage of the Safe Area Inset environment variables in CSS. This allows the body's background image to take up all of the space on the screen, where available. This is enabled from within the viewport meta tag, which then allows the page layout to be fully managed from CSS after that.
- Added a few fallbacks to the body's font family declaration. I noticed that system-ui wasn't working on Chrome OS, and I believe Firefox still doesn't have support for it yet. These additional fallbacks will allow the system font to appear correctly for more devices and browsers that don't yet support system-ui.

BenjaminAster reviewed (approved):
Thanks for taking the time to check out and improve my website!

I've never worked with the safe-area-inset environment variables, but I guess it makes sense to care about display cutouts in website design. I just  don't like it that the safe-area-inset environment variables don't have logical versions. [There's already an issue about that on the CSS working group's drafts](https://github.com/w3c/csswg-drafts/issues/6379).
In line 53, I'd prefer four different lines for padding-top, padding-right, padding-bottom and padding-left instead of one giant unreadable line.

About the font families: [All modern browsers (including Firefox) should support system-ui](https://caniuse.com/font-family-system-ui).
![image](https://user-images.githubusercontent.com/63414473/162896197-a5e81487-9242-45b3-9b0a-9a2eb724aa57.png)
I don't like other font families like Segoe UI and Roboto being before system-ui because if I change my operating system font I want the website to react to that. But if I have either Segoe UI or Roboto installed, my operating system's font will be ignored. I'm okay with providing a sans-serif fallback for ChromeOS, so the font-family should simply be `system-ui, sans-serif`.

I'll approve your pull request for now, and then change and push the mentioned things.

BenjaminAster commented:
One more thing: As you might have seen, I have a small server side rendering script written in [Deno](https://deno.land) (TypeScript) that fetches the descriptions and icons of my projects from benjaminaster.com. So if you want to change some HTML, you should change it in ./ssr/template.html and then re-run the ssr.deno.ts script. This will then overwrite index.html with the changes you made in template.html.

Offroaders123 commented:
> I just don't like it that the safe-area-inset environment variables don't have logical versions.

I thought the same thing later once thinking about it a little more, this would be great to have.

> In line 53, I'd prefer four different lines for padding-top, padding-right, padding-bottom and padding-left instead of one giant unreadable line.

Yeah, that would definitely make more sense, haha. My definition was really packed.

> All modern browsers (including Firefox) should support system-ui.

Glad to hear about this! Should've tested this again myself I didn't hear that it was implemented, looks like it recently got added only a few months ago, last September.

Offroaders123 commented:
Oh yeah, and I forgot about a question I had while working on the pull request:

Is the custom scroll behavior for the `<body>` with `position: fixed` present to add a similar appearance to `background-attachment: fixed` for Safari? Just curious what those were for, and whether that was the cause for including those. Smart workaround by the way, if so! Safari can be a little challenging at times, haha.

BenjaminAster commented:
> Is the custom scroll behavior for the `<body>` with `position: fixed` present to add a similar appearance to `background-attachment: fixed` for Safari?

Well actually... no, it isn't. It has to do with the background-image showing behind the scrollbar on Blink (and WebKit?) based desktop browsers. Normally, the `<body>` is displayed _next to_ the scrollbar. So if there's a background-image on the `<body>`, the scrollbar's background appears in a solid color. What I did now is set the `:root`'s `overflow` to `hidden` and gave the `<body>` a `position: fixed` and `inset: 0`. That way, the body element is fixed to the viewport, so the dimensions of the `<body>` are equivalent to the viewport dimensions. All the content now overflows _inside_ of the `<body>` which leads to the scrollbar being also _inside_ the body. This gives the desired result of the background-image appearing behind the scollbar. I've played around with that for a way too long time without the desired result until I realized that you also have to give the `<body>` an `overflow-x: hidden` to make that work (at least in Blink, and probably also in WebKit). This makes absolutely no sense to me, but that way it works.

As for the background-attachment: Thanks for letting me know that [WebKit doesn't support that](https://caniuse.com/background-attachment)!
![image](https://user-images.githubusercontent.com/63414473/162938547-13e17fbb-cfe3-4e36-b8b0-3b0d0f99a468.png)
I can't test my websites on WebKit because I don't have any Apple device. But if I understand that correctly, the background-image's position is actually displayed with a fixed position on WebKit? That would be a nice coincidence, as I tried to solve one problem and accidentally solved another problem that was not known to me at all :)

Offroaders123 commented:
> But if I understand that correctly, the background-image's position is actually displayed with a fixed position on WebKit? That would be a nice coincidence, as I tried to solve one problem and accidentally solved another problem that was not known to me at all :)

Sounds exactly like this! Can't believe it worked out that way.

I was debating whether the `background-image` behind the scrollbar could've been related to the `position: fixed` behavior, interesting it ended up being that. Caught me by surprise! Great design choice for that too. I thought it could have just been a side effect for the issue mentioned above, while it was actually the other way around.

Offroaders123 commented:
Alright last comment, sorry 😆

I found [a very helpful Apple Shortcut](https://www.macstories.net/ios/apple-frames-2-0-faster-lighter-and-featuring-support-for-iphone-13-ipad-mini-ipad-10-2-imac-24-macbook-air-and-multiple-languages/) that allows you to add device frames around your own screenshots, so I thought I'd do the same for my original screenshots above.

I use the iPhone 12 mini, and the device frame looks like the 13 mini instead, so the notch in the picture is a little smaller in width than my iPhone's in person, but it's close enough to get the job done.

### Viewport Fit: Default

![Viewport Fit Default](https://user-images.githubusercontent.com/65947371/162948760-eb841b01-f238-4aaa-b96c-146951d3a373.png)

### Viewport Fit: Cover

![Viewport Fit Cover](https://user-images.githubusercontent.com/65947371/162948801-5f1f0f85-5040-4e89-94b9-1926c5c82bdf.png)

BenjaminAster commented:
Hey @Offroaders123, look what I found in [Safari 15.5's release notes](https://webkit.org/blog/12669/new-webkit-features-in-safari-15-5/):
![CSS Fixed scrolling background-attachement: fixed](https://user-images.githubusercontent.com/63414473/169761439-6faa33fb-7ea4-41ea-adba-27b90c30adaa.png)

Offroaders123 commented:
Started testing around with it in Safari, and thought it still wasn't working. Realized I haven't updated to 15.5 yet, haha.
Installing the new update, then I'll see how support for it is after that!

Offroaders123 commented:
Maybe I could be implementing it wrong, but it still doesn't seem to be working as intended in 15.5 😕

BenjaminAster commented:
Hmmm, that's interesting! 🤔 As I said I don't have any access to a WebKit browser, so I can't really try it out...
